### PR TITLE
TINKERPOP-980: added sysv and systemd init scripts

### PR DESCRIPTION
**Pull request by robertdale**
I don't know if you want all of this in bin/.  I could definitely see the systemd gremlin-server.service moving out since it's more of a template but I don't know where else to put it.

### CODE REVIEW

**Comment by PommeVerte** [comment]
Should we add upstart to this PR? I'm not sure if the systemd covers older versions of ubuntu. Namely 14.04LTS ?

**Comment by PommeVerte** [comment]
Just confirming that although ubuntu 14.04 has `systemd` it does not have `systemd-as-init` . So an upstart would be a good idea and should be relatively easy since this PR seems to be thorough.

Question is, do we make a separate PR for this or not? 

Also maybe we should add some more documentation around this (instead of only having the upgrade docs)? Like in the gremlin-server section.

**Comment by spmallette** [comment]
I like the suggestion @PommeVerte made to have some docs in the gremlin-server section somewhere. Then you could just reference them from the upgrade docs.

I don't know what an upstart script entails so i'm not sure if a separate body of work for it is required or not.

Finally, not sure where this should be in the repo. Is it customary to find these in a `bin/` directory?

**Comment by robertdale** [comment]
Upstart looks simple enough. I don't know if the one @dkuppitz created, referenced in TINKERPOP-980 , is still viable. But I can't test. Any volunteers?

I would have to say that it's not customary for systemd/upstart templates to be in `bin/`.  They're really not binaries, more like configuration for those apps. Maybe create an `extras/` dir?

The initd script is a real, executable script with `start, stop, status`.  I didn't want to modify `gremlin-server.sh` since it has some functionality over a pure init script like install deps, run in foreground, pass conf file.  However, it also has its short-comings like it can only be run in the install dir.  The initd script can be run from anywhere and also can be directly referenced from initd.  That said, it's not unusual for the main server script to also be the initd script.

My preference would be to merge the two. But there *should* be a breaking change in that executing `gremlin-server.sh` would show help instead of running in the foreground.  Would probably add a new option to run in foreground, `start -f`.   Passing a conf file can still work, with or without start. `-i` can still work.

**Comment by robertdale** [comment]
Maybe if the script detects that it's called via link with no parameters, then it's in initd mode and displays help. If the script is called directly with no parameters, then it just runs in foreground.   Then there's no breaking change.

**Comment by robertdale** [comment]
- `gremlin-server.sh` _is_ the init script.  This makes it more inline with other Apache servers like Tomcat, ActiveMQ, solr, etc.
- Moved how to install "As A Service" from Upgrade to Gremlin Server Configuration per @PommeVerte 
- Also added parameters and environment config to Server Configuration section
- Deprecated `-i` in favor of `install`
- Starting `gremlin-server.sh` without parameters is a breaking change. Amazingly, it's not ever actually used anywhere without parameters.

File: docs/src/upgrade/release-3.2.x-incubating.asciidoc [hunk 1]
```diff
@@ -152,6 +152,15 @@ there is no need to do that manually.
...
 See: link:https://issues.apache.org/jira/browse/TINKERPOP-790[TINKERPOP-790]
 
+Gremlin-server.sh and Init Scripts
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+`gremlin-server.sh` is now also an init script and can no longer be started without parameters. To start it in the foreground, please use the `console` parameter. Also, it will still start in the foreground when provided a yaml configuration file.
+
+How to install as a service has been added to the reference docs.
+
+See: link:https://issues.apache.org/jira/browse/TINKERPOP-980[TINKERPOP-980]
```
Comment (spmallette) [inline docs/src/upgrade/release-3.2.x-incubating.asciidoc]: I would add a link to the section in the reference docs here in the "See:" section like was done [here](https://github.com/apache/tinkerpop/blob/master/docs/src/upgrade/release-3.2.x-incubating.asciidoc#graphson-20)

**Comment by spmallette** [comment]
wow - @robertdale - this is a really nice PR. I added the "breaking" label to the JIRA ticket as this would prevent gremlin-server.sh from working as it did before. I think I'm ok with that, but we should probably have a DISCUSS on the dev list about that just to bring it to everyone's attention to make sure there are no objections. Do you mind starting that thread for us - otherwise I will kick one off?

As for the PR itself, I can test this manually, but will have to rely on others for code review. gremlin-server.sh is looking pretty legit!

**Comment by robertdale** [comment]
Cool. I was wondering how to link. Didn't think it was absolute.  I hope it works. Tried testing with `./docker/build.sh -d` but it hangs at installing plugins `* hadoop-gremlin ...`.

**Comment by spmallette** [comment]
It occurred to me that it would seem there must be a place that gremlin-server.sh is used in doc generation - and here it is: 

https://github.com/apache/tinkerpop/blob/8ab682fef3e3da12904cb0d825abf4cc1ba8235a/docs/preprocessor/preprocess.sh#L84

I assume that command will no longer work with these changes and a fix is needed there? @dkuppitz any other secret areas you know about where gremlin-server.sh is used?

**Comment by robertdale** [comment]
That should still work.  What won't work is an invocation of just "gremlin-server.sh" without parameters.  I didn't find any in the code base nor in the docs (which is highly convenient).

**Comment by spmallette** [comment]
oh - right - you had said that. sorry

**Comment by robertdale** [comment]
I just tested with backgrounding the job and killing that.  That also still works.

**Comment by PommeVerte** [comment]
I'm currently on some ressource intensive tasks so it's hard for me to bounce back and forth between projects but I would like to give this a run on ubuntu 14.04. I suspect that sysV might work out of the box with a simple change in commands. 
Will give this a spin asap.

**Comment by robertdale** [comment]
@PommeVerte  do you mind if we move the ubuntu script to a separate pull request so this one can move forward?

**Comment by spmallette** [comment]
I believe @PommeVerte is on a month long holiday at this point. imo, i think we could move the ubuntu script off to a different ticket as you have a lot of work in place here. I would just rebase on master, move your changelog/upgrade docs stuff to be under 3.3.0 and we can move to vote. 

And of course, now that you're a committer @robertdale you should VOTE on your own stuff for a free +1 :)

**Comment by robertdale** [comment]
Rebased. Retested.

VOTE +1

**Comment by spmallette** [comment]
Tested all the command line options and all work nicely. Good documentation.

It did occur to me that we don't have analogous features for windows in that the gremlin-server.sh does not behave like the gremlin-server.bat. I'd suggest two things:

1.  Maybe just add a "WARNING:" to the docs to mention that discrepancy and that gremlin-server.bat always starts Gremlin Server in the foreground.
2. Change gremlin-server.bat to use "install" instead of "-i"

Other than that, looks good to me:

VOTE +1

**Comment by robertdale** [comment]
@spmallette I pushed the windows changes. can you review?

**Comment by spmallette** [comment]
LGTM

**Comment by spmallette** [comment]
@PommeVerte do you think you might have some free time soon to review this and vote?

File: docs/src/reference/gremlin-applications.asciidoc [hunk 2]
```diff
@@ -1482,6 +1513,62 @@ and standard deviation evaluation times, as well as the 75th, 95th, 98th, 99th a
...
+
+# Or call directly
+/etc/init.d/gremlin-server restart
+
+----
+
+Systemd
++++++++
+
+To install, copy the this service template to /etc/systemd/system/gremlin.service
```
Comment (dkuppitz) [inline docs/src/reference/gremlin-applications.asciidoc]: Minor typo: "the this"

File: docs/src/reference/gremlin-applications.asciidoc [hunk 3]
```diff
@@ -1482,6 +1513,62 @@ and standard deviation evaluation times, as well as the 75th, 95th, 98th, 99th a
...
+and update the paths `/path/to/apache-tinkerpop-gremlin-server` with the actual install path of gremlin server.
```
Comment (dkuppitz) [inline docs/src/reference/gremlin-applications.asciidoc]: gremlin server => Gremlin Server

File: docs/src/reference/gremlin-applications.asciidoc [hunk 4]
```diff
@@ -1482,6 +1513,62 @@ and standard deviation evaluation times, as well as the 75th, 95th, 98th, 99th a
...
+
+[source,bash]
+----
+[Unit]
+Description=Apache Tinkerpop Gremlin server daemon
```
Comment (dkuppitz) [inline docs/src/reference/gremlin-applications.asciidoc]: Tinkerpop => TinkerPop
Gremlin server => Gremlin Server

File: docs/src/upgrade/release-3.3.x.asciidoc [hunk 5]
```diff
@@ -31,3 +31,20 @@ Please see the link:https://github.com/apache/tinkerpop/blob/3.3.3/CHANGELOG.asc
...
 Upgrading for Users
 ~~~~~~~~~~~~~~~~~~~
+
+
+Gremlin-server.sh and Init Scripts
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+`gremlin-server.sh` is now also an init script and can no longer be started without parameters. To start it in the foreground with defaults like previous usage, please use the `console` parameter. Also, `gremlin-server.sh` will continue to start in the foreground when provided a yaml configuration file.
+
+How to install as a service has been added to the link:http://tinkerpop.apache.org/docs/3.2.3/reference/#_as_a_service[Reference Documentation - As A Service].
```
Comment (dkuppitz) [inline docs/src/upgrade/release-3.3.x.asciidoc]: The version in this (and the following) link should be 3.3.3.

File: gremlin-server/src/main/bin/gremlin-server.conf [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+# initd: debug the initd script
+#DEBUG=1
+
+# initd: gremlin server install location
```
Comment (dkuppitz) [inline gremlin-server/src/main/bin/gremlin-server.conf]: I know, it's only a comment, but....

gremlin server => Gremlin Server

File: gremlin-server/src/main/bin/gremlin-server.conf [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+#GREMLIN_HOME=
+
+# initd: Yaml file passed to gremlin server. Path can be absolute or relative to GREMLIN_HOME e.g. conf/gremlin-server.yaml
```
Comment (dkuppitz) [inline gremlin-server/src/main/bin/gremlin-server.conf]: same as above

**Comment by dkuppitz** [comment]
I can't test Systemd, but I've installed Gremlin Server as a service and that worked like a charm. Manual starts / restarts / stops /status requests worked as expected. Killing the process didn't lead to any unexpected behavior.

Next I installed the service as a real system service (I may have overlooked it, but I believe this is not mentioned in the docs):

```
daniel@cube ~ $ sudo update-rc.d -f gremlin-server defaults
 Adding system startup for /etc/init.d/gremlin-server ...
   /etc/rc0.d/K20gremlin-server -> ../init.d/gremlin-server
   /etc/rc1.d/K20gremlin-server -> ../init.d/gremlin-server
   /etc/rc6.d/K20gremlin-server -> ../init.d/gremlin-server
   /etc/rc2.d/S20gremlin-server -> ../init.d/gremlin-server
   /etc/rc3.d/S20gremlin-server -> ../init.d/gremlin-server
   /etc/rc4.d/S20gremlin-server -> ../init.d/gremlin-server
   /etc/rc5.d/S20gremlin-server -> ../init.d/gremlin-server
daniel@cube ~ $ sudo update-rc.d -f gremlin-server enable
update-rc.d: warning:  start runlevel arguments (none) do not match gremlin-server Default-Start values (2 3 4 5)
update-rc.d: warning:  stop runlevel arguments (none) do not match gremlin-server Default-Stop values (0 1 6)
 Enabling system startup links for /etc/init.d/gremlin-server ...
 Removing any system startup links for /etc/init.d/gremlin-server ...
   /etc/rc0.d/K20gremlin-server
   /etc/rc1.d/K20gremlin-server
   /etc/rc2.d/S20gremlin-server
   /etc/rc3.d/S20gremlin-server
   /etc/rc4.d/S20gremlin-server
   /etc/rc5.d/S20gremlin-server
   /etc/rc6.d/K20gremlin-server
 Adding system startup for /etc/init.d/gremlin-server ...
   /etc/rc0.d/K20gremlin-server -> ../init.d/gremlin-server
   /etc/rc1.d/K20gremlin-server -> ../init.d/gremlin-server
   /etc/rc6.d/K20gremlin-server -> ../init.d/gremlin-server
   /etc/rc2.d/S20gremlin-server -> ../init.d/gremlin-server
   /etc/rc3.d/S20gremlin-server -> ../init.d/gremlin-server
   /etc/rc4.d/S20gremlin-server -> ../init.d/gremlin-server
   /etc/rc5.d/S20gremlin-server -> ../init.d/gremlin-server
```

Then I rebooted my system and Gremlin Server started right away. Great! Finally I removed the system service:

```
daniel@cube ~ $ sudo update-rc.d gremlin-server disable
update-rc.d: warning:  start runlevel arguments (none) do not match gremlin-server Default-Start values (2 3 4 5)
update-rc.d: warning:  stop runlevel arguments (none) do not match gremlin-server Default-Stop values (0 1 6)
 Disabling system startup links for /etc/init.d/gremlin-server ...
 Removing any system startup links for /etc/init.d/gremlin-server ...
   /etc/rc0.d/K20gremlin-server
   /etc/rc1.d/K20gremlin-server
   /etc/rc2.d/S20gremlin-server
   /etc/rc3.d/S20gremlin-server
   /etc/rc4.d/S20gremlin-server
   /etc/rc5.d/S20gremlin-server
   /etc/rc6.d/K20gremlin-server
 Adding system startup for /etc/init.d/gremlin-server ...
   /etc/rc0.d/K20gremlin-server -> ../init.d/gremlin-server
   /etc/rc1.d/K20gremlin-server -> ../init.d/gremlin-server
   /etc/rc6.d/K20gremlin-server -> ../init.d/gremlin-server
   /etc/rc2.d/K80gremlin-server -> ../init.d/gremlin-server
   /etc/rc3.d/K80gremlin-server -> ../init.d/gremlin-server
   /etc/rc4.d/K80gremlin-server -> ../init.d/gremlin-server
   /etc/rc5.d/K80gremlin-server -> ../init.d/gremlin-server
daniel@cube ~ $ sudo update-rc.d -f gremlin-server remove
 Removing any system startup links for /etc/init.d/gremlin-server ...
   /etc/rc0.d/K20gremlin-server
   /etc/rc1.d/K20gremlin-server
   /etc/rc2.d/K80gremlin-server
   /etc/rc3.d/K80gremlin-server
   /etc/rc4.d/K80gremlin-server
   /etc/rc5.d/K80gremlin-server
   /etc/rc6.d/K20gremlin-server
```

That's it. No surprises, everything's working well. Nothing to add here, except ...

VOTE: +1

**Pull request merged**